### PR TITLE
Converting to the C++ librdkafka API 

### PR DESCRIPTION
**Pull request by rerpha**
### Description of work

Converting the Consumer and Producer to use the C++ librdkafka API. 

System tests have been modified to create and run the Kafka, Zookeeper and Softioc containers before starting the forwarder and persisting throughout the test session. This is so that the forwarder can start as the config topic should already exist on Kafka, and also it's a more likely use case. This should also make the tests run quicker as there is only one container each test fixture to be stopped and removed (the forwarder container) 

### Issue

DM-1135

branch `longrunning_systest` should also be deleted when this PR is merged

### Acceptance Criteria

Now uses one callback for all producer delivery reports
Two Callbacks for Events (Error, stats, log) - one cons, one prod

Also various refactoring to make the code clearer and easier to use. 

### Unit Tests

None.

### Other

Code review should be set up for this PR after tests have been added. 

---

## Code Review (To be filled in by the reviewer only)

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [x] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

**Comment by rerpha** [comment]
Seems to be working, I think a code review should be set up for this 

also need to add some more tests

File: src/KafkaW/MetadataException.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <exception>
+#include <string>
+
+class MetadataException : public std::exception {
```
Comment (mattclarke) [inline src/KafkaW/MetadataException.h]: This class is not actually used because the code no longer check the metadata.

**Comment by matthew-d-jones** [comment]
Since these are big changes I think it would be good to have a long running test that could be left going overnight. Perhaps by updating a PV with consecutive values and running a consumer which logs any deviation from the next expected value. Memory use should be (manually) checked at the end of the test. The test could be scripted in the system tests but be disabled by default.

**Comment by rerpha** [comment]
Ran the long-running low freq update test last night, no pv updates were missing on the topic (in 8.5 hours) 

going to run again tonight with memory usage logging, to see if there's a significant memory leak.

**Comment by rerpha** [comment]
Something a bit odd is going on. The test runner seems to sometimes segfault, for no apparent reason. I'd imagine its either a memory leak in the forwarder (unlikely as its containerised) or pytest/docker itself?

File: src/KafkaW/ProducerTopic.cpp [hunk 2]
```diff
@@ -44,59 +25,67 @@ ProducerTopic::ProducerTopic(ProducerTopic &&x) noexcept {
...
   MsgPtr->finalize();
-  unique_ptr<Producer::Msg> Msg(MsgPtr);
+  std::unique_ptr<ProducerMessage> Msg(MsgPtr);
   return produce(Msg);
 }
 
-int ProducerTopic::produce(unique_ptr<Producer::Msg> &Msg) {
-  int x;
-  int32_t partition = RD_KAFKA_PARTITION_UA;
+int ProducerTopic::produce(std::unique_ptr<ProducerMessage> &Msg) {
```
Comment (matthew-d-jones) [inline src/KafkaW/ProducerTopic.cpp]: This could do with doxygen comment

File: src/KafkaW/ProducerTopic.cpp [hunk 3]
```diff
@@ -44,59 +25,67 @@ ProducerTopic::ProducerTopic(ProducerTopic &&x) noexcept {
...
   }
 };
 
+/// NB this copies the provided data - so use only for low volume publishing
+/// \param MsgData Pointer to the data to publish
+/// \param MsgSize Size of the data to publish
+/// \return 0 if message is successfully passed to RdKafka to be published, 1
+/// otherwise
 int ProducerTopic::produce(unsigned char *MsgData, size_t MsgSize) {
   auto MsgPtr = new Msg_;
```
Comment (matthew-d-jones) [inline src/KafkaW/ProducerTopic.cpp]: `make_unique`

File: src/KafkaW/ConsumerMessage.h [hunk 4]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include <cstdint>
+#include <cstdlib>
+
+namespace KafkaW {
+
+enum class PollStatus {
+  Msg,
```
Comment (rerpha) [inline src/KafkaW/ConsumerMessage.h]: Full names

File: src/Kafka.h [hunk 5]
```diff
@@ -15,7 +15,6 @@
 
 #include "FlatbufferMessage.h"
 #include "KafkaW/KafkaW.h"
-#include <librdkafka/rdkafka.h>
 
 namespace Forwarder {
```
Comment (matthew-d-jones) [inline src/Kafka.h]: Remove `using sptr`

File: src/KafkaW/Consumer.cpp [hunk 6]
```diff
@@ -1,198 +1,126 @@
...
-    LOG(Sev::Debug, "rd_kafka_consumer_close");
-    rd_kafka_consumer_close(RdKafka);
-    LOG(Sev::Debug, "rd_kafka_destroy");
-    rd_kafka_destroy(RdKafka);
-    RdKafka = nullptr;
-  }
-  if (PartitionList) {
-    rd_kafka_topic_partition_list_destroy(PartitionList);
-    PartitionList = nullptr;
+std::unique_ptr<RdKafka::Metadata> Consumer::queryMetadata() {
```
Comment (SkyToGround) [inline src/KafkaW/Consumer.cpp]: This member function can be improved as suggested in the comment.

File: src/KafkaW/Consumer.cpp [hunk 7]
```diff
@@ -1,198 +1,126 @@
...
-  for (int i1 = 0; i1 < PartitionList->cnt; ++i1) {
-    auto &Partition = PartitionList->elems[i1];
-    LOG(Sev::Debug, "   {}  {}  {}", Partition.topic, Partition.partition,
-        Partition.offset);
+Consumer::~Consumer() {
+  LOG(Sev::Debug, "~Consumer()");
+  if (KafkaConsumer != nullptr) {
+    LOG(Sev::Debug, "Close the consumer");
+    KafkaConsumer->close();
+    RdKafka::wait_destroyed(5000);
```
Comment (SkyToGround) [inline src/KafkaW/Consumer.cpp]:
```suggestion
```
This line is no longer required according to [the documentation](https://github.com/edenhill/librdkafka/wiki/Proper-termination-sequence).

File: src/KafkaW/Consumer.cpp [hunk 8]
```diff
@@ -1,198 +1,126 @@
...
-#define KERR(RK, ERR)                                                          \
-  if ((ERR) != 0) {                                                            \
-    LOG(Sev::Error, "Kafka {}  error: {}, {}, {}", rd_kafka_name(RK), ERR,     \
-        rd_kafka_err2name((rd_kafka_resp_err_t)ERR),                           \
-        rd_kafka_err2str((rd_kafka_resp_err_t)ERR));                           \
+Consumer::Consumer(BrokerSettings &BrokerSettings)
+    : ConsumerBrokerSettings(std::move(BrokerSettings)) {
+  std::string ErrorString;
+
+  Conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
```
Comment (SkyToGround) [inline src/KafkaW/Consumer.cpp]:
```suggestion
  Conf = std::unique_ptr<RdKafka::Conf>(RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL));
```
Should not be raw pointer.

File: src/KafkaW/Consumer.cpp [hunk 9]
```diff
@@ -1,198 +1,126 @@
...
+  Conf->set("event_cb", &EventCallback, ErrorString);
+  Conf->set("metadata.broker.list", ConsumerBrokerSettings.Address,
+            ErrorString);
+  Conf->set("group.id",
+            fmt::format("forwarder-command-listener--pid{}", getpid()),
+            ErrorString);
+  ConsumerBrokerSettings.apply(Conf);
+
+  KafkaConsumer = std::shared_ptr<RdKafka::KafkaConsumer>(
+      RdKafka::KafkaConsumer::create(Conf, ErrorString));
```
Comment (SkyToGround) [inline src/KafkaW/Consumer.cpp]:
```suggestion
      RdKafka::KafkaConsumer::create(Conf.get(), ErrorString));
```
Not raw pointer.

File: src/KafkaW/Consumer.cpp [hunk 10]
```diff
@@ -1,198 +1,126 @@
...
+  std::string ErrorString;
+
+  Conf = RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL);
+  Conf->set("event_cb", &EventCallback, ErrorString);
+  Conf->set("metadata.broker.list", ConsumerBrokerSettings.Address,
+            ErrorString);
+  Conf->set("group.id",
+            fmt::format("forwarder-command-listener--pid{}", getpid()),
+            ErrorString);
+  ConsumerBrokerSettings.apply(Conf);
```
Comment (SkyToGround) [inline src/KafkaW/Consumer.cpp]:
```suggestion
  ConsumerBrokerSettings.apply(Conf.get());
```
Not raw pointer.

File: src/KafkaW/Consumer.h [hunk 11]
```diff
@@ -1,77 +1,38 @@
...
-      on_rebalance_start;
-  rd_kafka_t *RdKafka = nullptr;
+  void addTopic(const std::string &Topic) override;
+  std::unique_ptr<ConsumerMessage> poll() override;
 
+protected:
+    std::shared_ptr<RdKafka::KafkaConsumer> KafkaConsumer;
+    virtual std::vector<int32_t> getTopicPartitionNumbers(const std::string &Topic);
 private:
+    RdKafka::Conf* Conf;
```
Comment (SkyToGround) [inline src/KafkaW/Consumer.h]:
```suggestion
    std::unique_ptr<RdKafka::Conf> Conf;
```
Do not use raw pointer.

File: src/KafkaW/Consumer.cpp [hunk 12]
```diff
@@ -1,198 +1,126 @@
...
-    LOG(Sev::Debug, "   {}  {}  {}", Partition.topic, Partition.partition,
-        Partition.offset);
+Consumer::~Consumer() {
+  LOG(Sev::Debug, "~Consumer()");
+  if (KafkaConsumer != nullptr) {
+    LOG(Sev::Debug, "Close the consumer");
+    KafkaConsumer->close();
+    RdKafka::wait_destroyed(5000);
   }
+  delete Conf;
```
Comment (SkyToGround) [inline src/KafkaW/Consumer.cpp]:
```suggestion
```
Not raw pointer.

File: src/KafkaW/Consumer.cpp [hunk 13]
```diff
@@ -1,198 +1,126 @@
...
-    if (CallbackERR != RD_KAFKA_RESP_ERR_NO_ERROR) {
-      LOG(Sev::Warning, "rebalance error: {}  {}",
-          rd_kafka_err2name(CallbackERR), rd_kafka_err2str(CallbackERR));
-    }
-    break;
+const RdKafka::TopicMetadata*
+Consumer::findTopic(const std::string &Topic) {
+  auto MetadataPtr = queryMetadata();
+  auto Topics = MetadataPtr->topics();
+  auto Iterator = std::find_if(Topics->cbegin(), Topics->cend(),
```
Comment (SkyToGround) [inline src/KafkaW/Consumer.cpp]: 👍

File: src/KafkaW/Producer.cpp [hunk 14]
```diff
@@ -5,95 +5,17 @@ namespace KafkaW {
 
 static std::atomic<int> g_kafka_producer_instance_count;
```
Comment (SkyToGround) [inline src/KafkaW/Producer.cpp]: This name sticks out like a sore thumb, please LLVM:ify it.

File: src/KafkaW/Producer.cpp [hunk 15]
```diff
@@ -5,95 +5,17 @@ namespace KafkaW {
...
-  auto Self = reinterpret_cast<Producer *>(Opaque);
-  LOG(Sev::Debug, "IID: {}  INFO cb_throttle  broker_id: {}  broker_name: {}  "
-                  "throttle_time_ms: {}",
-      Self->id, Broker_id, Name, Throttle_time_ms);
-}
-
 Producer::~Producer() {
   LOG(Sev::Debug, "~Producer");
-  if (RdKafkaPtr) {
+  if (ProducerPtr) {
```
Comment (SkyToGround) [inline src/KafkaW/Producer.cpp]:
```suggestion
  if (ProducerPtr != nullptr) {
```

File: src/KafkaW/Producer.cpp [hunk 16]
```diff
@@ -5,95 +5,17 @@ namespace KafkaW {
...
     int TimeoutMS = 1;
     int OutQueueLength = 0;
     while (true) {
```
Comment (SkyToGround) [inline src/KafkaW/Producer.cpp]: I do not like this while statement and I do not like how timeouts are handled handled here. It should be a simple thing to write something that is easier to understand.

File: src/KafkaW/ProducerTopic.cpp [hunk 17]
```diff
@@ -44,59 +26,67 @@ ProducerTopic::ProducerTopic(ProducerTopic &&x) noexcept {
   std::swap(DoCopyMsg, x.DoCopyMsg);
 }
 
-struct Msg_ : public Producer::Msg {
-  vector<unsigned char> v;
+struct Msg_ : public ProducerMessage {
+  std::vector<unsigned char> v;
```
Comment (SkyToGround) [inline src/KafkaW/ProducerTopic.cpp]: It is not a problem as such but I do find it slightly weird that you are using a `std:vector` here when there are several other (in my opinion) more obvious options.

**Comment by rerpha** [comment]
it works!

**Pull request merged**